exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ExtendedOptionIterator.java

```diff
@@ -0,0 +1,152 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.store.sys;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Arrays;
+
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Lists;
+import org.apache.drill.exec.ops.FragmentContext;
+import org.apache.drill.exec.server.options.OptionManager;
+import org.apache.drill.exec.server.options.OptionValue;
+import org.apache.drill.exec.server.options.OptionValue.Kind;
+import org.apache.drill.exec.server.options.OptionValue.OptionType;
+import org.apache.drill.exec.server.options.SystemOptionManager;
+/*
+ * To extend the original Option iterator to hide the implementation details and to return the row
+ * which takes precedence over others for an option.This is done by examining the type as the precendence
+ * order is session - system - default.All the values are represented as String instead of having multiple
+ * columns and the datatype is provided for type details.
+ */
+public class ExtendedOptionIterator implements Iterator<Object> {
+//  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(OptionIterator.class);
+
+  private final OptionManager fragmentOptions;
+  private final Iterator<OptionValue> mergedOptions;
+
+  public ExtendedOptionIterator(FragmentContext context) {
+    fragmentOptions = context.getOptions();
+    final Iterator<OptionValue> optionList;
+    mergedOptions = sortOptions(fragmentOptions.iterator());
+  }
+  /*
+    Sort options according to name and the type to remove the redundant rows
+    for the same option based on the type
+   */
+  public Iterator<OptionValue> sortOptions(Iterator<OptionValue> options)
+  {
+    List<OptionValue> values = Lists.newArrayList(options);
+    List<OptionValue> optionValues = Lists.newArrayList();
+    List<OptionValue> optionValueList = Lists.newArrayList();
+    OptionValue value;
+    OptionType type;
+    OptionValue[] temp = new OptionValue[3];
```

Comment:
I think this may be a bit easier to implement with a map. Add each "inner" option to a map keyed by value. If you find an existing entry, choose the one with higher precedence and replace the map value if needed.

Then, iterate over the map values to get the final objects. Iteration over the values can be done in each call to the `next()` method of your iterator to avoid the need to build up a list of extended value objects.